### PR TITLE
Check Samples and Libraries for Correct Format in Pipeline

### PR DESCRIPTION
**Pull request by ScottCarda-MS**
Expands on using the formatter from the command-line.
 - Allows folders to be given.
 - Adds the `--recursive` flag for specifying if you want the folders to be processed recursively. Without this, if given a folder it will only search the direct children of the folder.
 - Adds the `--write` flag for specifying that the formatter should overwrite unformatted Q# files with the format corrections. Without this flag, the formatter only checks if the Q# files are correctly formatted.
To use the command-line formatter run `cargo run --binqsc_formatter -- <args to formatter>`.

Adds a check to the CI pipeline to run the formatter against the libraries and samples, and stops a build if it finds files that are not formatted correctly.

### CODE REVIEW

**Comment by ScottCarda-MS** [comment]
> I think it would be advantageous to have a tighter integration with the project system here. What if, instead of `--recursive`, the formatter queried the project system for the project files and formatted them? And for single file mode, it would just take the name of a file to format? Then, we could orchestrate the CI with a basic script that calls the formatter on each individual file.
> 
> `prettier` doesn't operate this way, because the ecosystem lacks a canonical "project" definition, but `cargo fmt` does. The advantageous side effect here is that we have a clean CI story as well as a project-aware formatter.

I'm open to the idea of having an option for formatting a Q# project, but that doesn't cover the same use-case as the folder option, so I still wouldn't want to remove that option. It is very useful to be able to format a specific folder regardless of it is is associated with a project or even a subfolder of a project. This option allows the user a degree of flexibility that wouldn't be covered by a project option.

File: compiler/qsc_formatter/src/main.rs
Comment (DmitryVasilevsky) [inline compiler/qsc_formatter/src/main.rs:27]: Is "write" a common option in this case? If not, would you consider something a little more menacing, such as "overwrite"?

Comment (ScottCarda-MS) [reply compiler/qsc_formatter/src/main.rs:27]: `--write` or `-w` seems to be common for cli options where you want to specify that you are writing to a file. Prettier, for example, uses `--write`. I'm not sure I like `--overwrite` because the shorthand of that would be `-o` which I don't associate with writing. That's just a personal preference.

File: .github/workflows/ci.yml
Comment (minestarks) [inline .github/workflows/ci.yml:121]: IMO this should be a part of `build.py --check` as well, so that it runs by default locally and so that it's consistent with how/when we run `prettier` and `cargo fmt` today.

File: .github/workflows/ci.yml
Comment (minestarks) [inline .github/workflows/ci.yml:227]: Nittiest of nits: can you use a dash instead of underscore in the name to match the style of the other job names?

(And as I say this, I realize some of them use camelCase so we're already inconsistent :/ )

File: .github/workflows/ci.yml
Comment (minestarks) [inline .github/workflows/ci.yml:119]: What about katas?

File: .github/workflows/ci.yml
Comment (minestarks) [reply .github/workflows/ci.yml:121]: Oh and if it fails, can it say "please fix the formatting by running `<exact command I can just copy and paste>`"?

File: compiler/qsc_formatter/src/main.rs
Comment (minestarks) [inline compiler/qsc_formatter/src/main.rs:174]: Is there a good reason we'd fail to read files? Shouldn't the command fail if we had to skip files?

**Review by minestarks — CHANGES_REQUESTED** [review]

File: compiler/qsc_formatter/src/main.rs
Comment (ScottCarda-MS) [reply compiler/qsc_formatter/src/main.rs:174]: It didn't seem very ergonomic to me to stop the whole formatting if one of the files encountered couldn't be read from due to permissions.

Comment (ScottCarda-MS) [reply compiler/qsc_formatter/src/main.rs:174]: Or do you mean the formatter should still run all the way through but return an error code?

File: .github/workflows/ci.yml
Comment (ScottCarda-MS) [reply .github/workflows/ci.yml]: done

File: .github/workflows/ci.yml
Comment (ScottCarda-MS) [reply .github/workflows/ci.yml:121]: I added it to the build.py --check step, but I'm not sure about having it print out a custom message if that step fails. I think it always just tells the user that the subprocess failed.

File: .github/workflows/ci.yml
Comment (ScottCarda-MS) [reply .github/workflows/ci.yml:119]: Discussed offline: katas will need a bit more work to integrated them into the formatting checks. This can be done in a future PR.

File: build.py
Comment (minestarks) [inline build.py:197]: - I think this code should be contained inside the same `if` statement above, so that it only runs if we're building any Rust code.
- Instead of `step_start` could you simply do a `print()`? Unfortunately `step_start`/`step_end` logging doesn't support nested pairs (the pert timer gets thrown off) - and this is missing a `step_end` anyways

File: compiler/qsc_formatter/src/main.rs
Comment (minestarks) [inline compiler/qsc_formatter/src/main.rs:19]: I think `clap` allows `Vec<>` arguments. If it's trivial to pass in multiple paths in the same command, maybe that would be nicer since we can avoid invoking the command twice in the build/ci scripts.

File: .github/workflows/ci.yml
Comment (minestarks) [reply .github/workflows/ci.yml:121]: Either way. If it makes the script really awkward then I can live without a custom message :)

File: build.py
Comment (ScottCarda-MS) [reply build.py:197]: yes - `print()` instead of `step_start()`, but why would we only want to check the format of the Q# code if we compile Rust?

File: compiler/qsc_formatter/src/main.rs
Comment (ScottCarda-MS) [reply compiler/qsc_formatter/src/main.rs:184]: done

File: .github/workflows/ci.yml
Comment (ScottCarda-MS) [reply .github/workflows/ci.yml]: I tried to add an instructive statement to the situations where we want the user to run the formatter with the `--write` option.

File: build.py
Comment (minestarks) [reply build.py]: > why would we only want to check the format of the Q# code if we compile Rust?

Warning, subjective answer ahead :)

Because if I'm running `./build.py` with specific flags that skip compiling the Rust bits (which I often do, when working on just the JS/Python bits), it would be an annoying extra step to build & run the Q# formatter.

Almost everything in this file is contained within these `if build_xyz` checks and I think it's good to follow that example, instead of running anything unconditionally. The idea is, if you're working on a specific component, you want to be able to build & check only those components. 

So in which cases _should_ we run the formatter? There's no ideal answer but these would be my expectations:

- if I'm building the library (`build_cli`)
- if I'm building the samples (`build_samples`)
- if I'm building the formatter itself (`build_cli` again)

Some common use cases and my expectations on whether the Q# formatter should run:

`./build.py` - yes (since no arguments means ALL)
`./build.py --npm --vscode` - no
`./build.py --samples` - yes
`./build.py --cli` - yes
`./build.py --wasm --npm --vscode`  - no


So maybe I'm changing my answer slightly: `if build_cli or build_samples` seems like the ideal condition to check.

File: compiler/qsc_formatter/src/main.rs
Comment (minestarks) [inline compiler/qsc_formatter/src/main.rs:209]: Love the detailed error handling, but I'd probably move all the success cases to the top to make it more immediately obvious what the success conditions are.

**Review by minestarks — APPROVED** [review]
Responded about `./build.py` flags - looks great otherwise, thanks! Please remember to check how this looks in CI one last time before merging

File: .github/workflows/ci.yml
Comment (minestarks) [inline .github/workflows/ci.yml:106]: Sorry, noticed one more thing! This is purely about aesthetics again, but this should live just below the other `Format` job so that the CI output looks more organized.

image

Comment (ScottCarda-MS) [reply .github/workflows/ci.yml:106]: I'm not entirely sure how this all works, but the reason I put it there was because the formatter needs to be built, so I put it after the build steps. Does that not have any impact?

File: build.py
Comment (ScottCarda-MS) [reply build.py]: The number of possible option combinations is a bit large, so I'll defer to you on this.

Comment (ScottCarda-MS) [reply build.py]: done, `if build_cli or build_samples`

File: compiler/qsc_formatter/src/main.rs
Comment (ScottCarda-MS) [reply compiler/qsc_formatter/src/main.rs]: done

File: .github/workflows/ci.yml
Comment (minestarks) [reply .github/workflows/ci.yml:106]: Oh no these jobs all run in parallel on different agents. I think it's just `cargo run` building your code.

Comment (ScottCarda-MS) [reply .github/workflows/ci.yml]: done

**Pull request merged**
